client/src/protocol.ts

```diff
@@ -38,3 +38,24 @@ export const UnexposeRequest = new RequestType<
   void,
   void
 >("elm/unexpose");
+
+export const FindTestsRequest = new RequestType<
+  IFindTestsParams,
+  IFindTestsResponse,
+  void
+>("elm/findTests");
+
+export interface IFindTestsParams {
+  workspaceRoot: URI;
+}
+
+export interface IFindTestsResponse {
+  suites?: TestSuite[];
+}
+
+export type TestSuite = {
+  label: string | string[];
+  tests?: TestSuite[];
+  file: string;
```

Comment:
> But there still seems to be weird behavior:
> 
> 1. Run all tests
> 2. enable auto run
> 3. change stuff in a test file
> 4. Some grey nodes appear, that won't go away (at least not with the auto run)
>    tested on elm review

Today's version seems to improve that.
Instead of re-loading tests at every change in test code, it now copied location information over between runs.

Over time, as the user works on test code, location information will get out of sync and users will need to reload manually. Let's see how that behaves.


A  p

Reply:
> Wondering if this should be URI too.

I have no opinion here, I could not figure out what the conventions in LSP where.
My impression was that sometimes strings transport Uri values ...

Reply:
Yeah, it shouldn't matter for now.